### DISCUSSION TITLE
How AWS Amplify is different from AWS SDK for Swift?

### DISCUSSION THREAD

**Discussion by parthp027** [awslabs/aws-sdk-swift#1776] [Q&A] [answered]
Can you please help me to identify a difference between [AWS Amplify](https://github.com/aws-amplify/amplify-swift) and [AWS SDK](https://github.com/awslabs/aws-sdk-swift) for a Swift?

**Comment by jbelkins** [comment, marked as answer]
Sure.  Here's the high-level difference between the two:

Amplify provides a set of tools that make it easy to create mobile & desktop apps, along with managing backend services hosted in AWS.  Amplify provides a lot of ready-made functionality like authentication & storage that you can use with just a little configuration.

AWS SDK for Swift is a set of tools for accessing all AWS services.  Rather than providing services that are ready-to-go and focused on the needs of application developers, AWS SDK for Swift provides access to everything AWS has to offer, and lets you talk directly to those services.  If you need to use AWS services that aren't supported by Amplify, the AWS SDK for Swift is how you would do that.

By the way, you don't have to choose one or the other - you can use them together.  You can use Amplify as the base of your app, then use AWS SDK for Swift for accessing AWS services that are beyond the scope of Amplify.  (Amplify-swift actually uses AWS SDK for Swift to perform some of its interactions with AWS services, so if you install amplify-swift you'll get AWS SDK for Swift installed as well). My general advice for a new app would be to start with Amplify to get essential services in place, then use the AWS SDK for Swift for AWS features that are beyond the scope of Amplify.

Hope that helps!

**Reply by parthp027** [reply]
Thank you, @jbelkins, for the detailed explanation. I understand from the documentation of the [AWS SDK](https://github.com/awslabs/aws-sdk-swift) that it offers services like S3, Lambda, DynamoDB, and more. However, I am struggling to find how authentication can be achieved without using AWS Amplify.

Could you please provide a direct link to an example or documentation that explains how to implement authentication using the AWS SDK?

**Reply by jbelkins** [reply]
We are actually working on a code example for authentication; I expect it to be ready in the next few days.  Onceit's posted, I'll link it here.

**Reply by parthp027** [reply]
Sounds good.

I am planning to commence development on my application shortly and anticipate integrating AWS services such as Cognito, DynamoDB, Lambda, S3, AppConfig, SNS, and Pinpoint. Would you recommend utilizing Amplify for this task, given its current support for these services, or would it be more beneficial to employ the AWS SDK directly? I only want to use one of them.

**Reply by jbelkins** [reply]
If starting both client & server from scratch, and Amplify's features will meet your most immediate needs, Amplify might be the best choice to help you get both client & server set up quicker.  If you already have fully-featured backend systems and just need your client to connect to them, using AWS SDK for Swift for specific functions might be the better choice.

And remember, you're not locked into one or the other.  If you eventually need AWS functions outside of Amplify's feature set, you can pull in AWS SDK for Swift to access those later.

**Reply by Takemichi0** [reply]
@jbelkins Hi! It’s been about two weeks, and if the sample codes are ready, could you kindly share them here? I would really appreciate it. Thank you in advance for your help.

**Reply by jbelkins** [reply]
Sorry for the delay.  We have an example [here](https://docs.aws.amazon.com/sdk-for-swift/latest/developer-guide/apple-integration.html) for using Sign In With Apple to obtain AWS credentials from AWS STS.

We have another sample in the works for obtaining credentials from Cognito as well.  We'll post a link here in Discussions once that is ready as well.

**Reply by ramink** [reply]
Thanks for the Sign In With Apple example! FWIW, I'm really looking forward to the Cognito example, as I haven't been able to make much progress on achieving this on my own. I'd be happy to see a draft version of the example! ;-)

**Reply by lon9man** [reply]
@jbelkins 
> I expect it to be ready in the next few days.

**~2 months already passed since 26 September**. 

Dear maintaners, it is pretty hard to understand current flow related to authentication using Amplify and other wrappers.
most of the developers need **SIMPLE UNDERSTANDABLE FLOW** to use SDK **WITHOUT** any additional wrappers like `Cognito, Amplify, Apple Sign-In, Amazon Sign-In and etc`.

almost all developers have **ONLY**:
1. mobile application
2. Amazon infrastructure
maybe additionally backend with API (PHP, NodeJS, ...) for some purpose (like user logging-in, ...)

### EXPECTED WORKFLOW:
mobile app needs to use Amazon infrastructure directly
**->** for this app needs some PERMISSIONS 
**->** for this (as i see) app needs temporary STS token. 
or maybe exist some other available SIMPLE logics without additional overhead. 

please GIVE US real examples for described above workflow. how much time it needs? years?
thanks!
